core/node/node_storage_init/src/external_node/revert.rs

```diff
@@ -0,0 +1,50 @@
+use anyhow::Context as _;
+use tokio::sync::watch;
+use zksync_block_reverter::BlockReverter;
+use zksync_dal::{ConnectionPool, Core};
+use zksync_reorg_detector::ReorgDetector;
+use zksync_types::L1BatchNumber;
+use zksync_web3_decl::client::{DynClient, L2};
+
+use crate::RevertStorage;
+
+#[derive(Debug)]
+pub struct ExternalNodeRevert {
+    pub client: Box<DynClient<L2>>,
+    pub pool: ConnectionPool<Core>,
+    pub reverter: BlockReverter,
+}
+
+#[async_trait::async_trait]
+impl RevertStorage for ExternalNodeRevert {
+    async fn revert_storage(
+        &self,
+        to_batch: L1BatchNumber,
+        _stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<()> {
+        tracing::info!("Reverting to l1 batch number {to_batch}");
+        self.reverter.roll_back(to_batch).await?;
+        tracing::info!("Revert successfully completed");
+        Ok(())
+    }
+
+    async fn is_revert_needed(&self) -> anyhow::Result<Option<L1BatchNumber>> {
+        let mut reorg_detector = ReorgDetector::new(self.client.clone(), self.pool.clone());
+        // We don't need this kind of granularity for detecting stop signal; it's expected to be treated
+        // by the caller.
```

Comment:
Can you explain what you mean by this comment? `run_once()` is cancel-aware to deal with transient errors in the L2 RPC client (e.g., no network). So, making `is_revert_needed()` not cancel-aware can degrade UX for the node operator (it could become unresponsive to SIGINTs).